lib/paperclip/lazy_thumbnail.rb

```diff
@@ -1,20 +1,50 @@
 # frozen_string_literal: true
 
 module Paperclip
-  class LazyThumbnail < Paperclip::Thumbnail
+  class LazyThumbnail < Paperclip::Processor
+    class PixelGeometryParser
+      def self.parse(current_geometry, pixels)
+        width  = Math.sqrt(pixels * (current_geometry.width.to_f / current_geometry.height)).round.to_i
+        height = Math.sqrt(pixels * (current_geometry.height.to_f / current_geometry.width)).round.to_i
+
+        Paperclip::Geometry.new(width, height)
+      end
+    end
+
+    def initialize(file, options = {}, attachment = nil)
+      super
+
+      @crop = options[:geometry].to_s[-1, 1] == '#'
+      @current_geometry = options.fetch(:file_geometry_parser, Geometry).from_file(@file)
+      @target_geometry = options[:pixels] ? PixelGeometryParser.parse(@current_geometry, options[:pixels]) : options.fetch(:string_geometry_parser, Geometry).parse(options[:geometry].to_s)
+      @format = options[:format]
+      @current_format = File.extname(@file.path)
+    end
+
     def make
-      return File.open(@file.path) unless needs_convert?
-
-      if options[:geometry]
-        min_side = [@current_geometry.width, @current_geometry.height].min.to_i
-        options[:geometry] = "#{min_side}x#{min_side}#" if @target_geometry.square? && min_side < @target_geometry.width
-      elsif options[:pixels]
-        width  = Math.sqrt(options[:pixels] * (@current_geometry.width.to_f / @current_geometry.height)).round.to_i
-        height = Math.sqrt(options[:pixels] * (@current_geometry.height.to_f / @current_geometry.width)).round.to_i
-        options[:geometry] = "#{width}x#{height}>"
+      source = File.open(@file.path)
+
+      return source unless needs_convert?
+
+      vips = ImageProcessing::Vips.source(source)
+
+      vips = if @crop
+               vips.resize_to_fill(@target_geometry.width, @target_geometry.height, sharpen: false, crop: :attention)
+             else
+               vips.resize_to_limit(@target_geometry.width, @target_geometry.height, sharpen: false)
+             end
+
+      vips = vips.custom do |image|
+        image.mutate do |mutable|
+          image.get_fields.each do |field|
+            mutable.remove!(field) unless field == 'icc-profile-data'
+          end
+        end
       end
 
-      Paperclip::Thumbnail.make(file, options, attachment)
+      vips.convert(@format)
+          .saver(interlace: true, quality: 90)
```

Comment:
Interlaced defaults to false. Is this recommended?

Reply:
Yes, this will make Vips output progressive JPEGs, which are slightly (a few percent) larger than normal JPEGs, but will start displaying data much quicker.

Reply:
Love that.. Shame something newer formats seem to be missing progressive atm. progressive AVIF seems to be on the way.